pkg/registry/apis/provisioning/jobs/queue.go

```diff
@@ -0,0 +1,184 @@
+		logger:   slog.Default().With("logger", "job-queue"),
+	}
+}
+
+type jobStore struct {
+	logger   *slog.Logger
+	capacity int
+	workers  []Worker
+
+	// All jobs
+	jobs []provisioning.Job
+	rv   int64 // updates whenever changed
+
+	mutex sync.RWMutex
+}
+
+// Register a worker (inline for now)
+func (s *jobStore) Register(worker Worker) {
+	s.mutex.Lock()
+	s.workers = append(s.workers, worker)
+	s.mutex.Unlock()
+}
+
+// Add implements JobQueue.
+func (s *jobStore) Add(ctx context.Context, job provisioning.Job) (string, error) {
+	if job.Namespace == "" {
+		return "", fmt.Errorf("expecting namespace")
+	}
+	_, ok := job.Labels["repository"]
+	if !ok {
+		return "", fmt.Errorf("expecting repository name label")
+	}
+	_, ok = job.Labels["repository.type"] // TODO: ideally just based on the action type
+	if !ok {
+		return "", fmt.Errorf("expecting repository name label")
+	}
+	if job.Spec.Action == "" {
+		return "", fmt.Errorf("missing spec.action")
+	}
+
+	// Only for add
+	if job.Status.State != "" {
+		return "", fmt.Errorf("must add jobs with empty state")
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	s.rv++
+	job.ResourceVersion = strconv.FormatInt(s.rv, 10)
+	job.Status.State = provisioning.JobStatePending
+	job.Labels["action"] = string(job.Spec.Action) // make it searchable
+	job.Labels["state"] = string(job.Status.State)
+	job.Name = "j" + uuid.NewString()
+	job.CreationTimestamp = metav1.NewTime(time.Now())
+
+	jobs := make([]provisioning.Job, 0, len(s.jobs)+2)
+	jobs = append(jobs, job)
+	for i, j := range s.jobs {
+		if i >= s.capacity {
+			break
+		}
+		jobs = append(jobs, j)
+	}
+
+	// For now, start a thread processing each job
+	go func() {
+		time.Sleep(time.Millisecond * 100)
+		s.drainPending()
+	}()
+
+	s.jobs = jobs // replace existing list
+	return job.Name, nil
+}
+
+// Reads the queue until no jobs remain
+func (s *jobStore) drainPending() {
+	var err error
+	for {
+		time.Sleep(time.Microsecond * 100)
+		ctx := context.Background()
+
+		job := s.Checkout(ctx, nil)
+		if job == nil {
+			return // done
+		}
+
+		var status *provisioning.JobStatus
+		var worker Worker
+		for _, w := range s.workers {
+			if w.Supports(ctx, job) {
+				worker = w
+				break
+			}
+		}
+		if worker == nil {
+			status = &provisioning.JobStatus{
+				State: provisioning.JobStateError,
+				Errors: []string{
+					"no registered worker supports this job",
+				},
+			}
+		} else {
+			status, err = worker.Process(ctx, *job)
+			if err != nil {
+				status = &provisioning.JobStatus{
+					State:  "error",
+					Errors: []string{err.Error()},
+				}
+			} else if status.State == "" {
+				status.State = provisioning.JobStateFinished
+			}
+		}
+
+		err = s.Complete(ctx, job.Namespace, job.Name, *status)
+		if err != nil {
+			s.logger.Error("error running job", "job", job.Name, "error", err)
+		}
+	}
+}
+
+// Checkout the next "pending" job
+func (s *jobStore) Checkout(ctx context.Context, query labels.Selector) *provisioning.Job {
```

Comment:
shouldn't this one be private?

Reply:
right now it does not matter, so maybe -- I was targeting a pattern where we have an async worker that can checkout a job then complete it.

But this hacks the worker management directly into the queue, so does not matter (and won't distribute across multiple machines!). Lets leave it for now and it will evolve as the worker gets extracted from the queue itself